### PR TITLE
Predicate Crafting Entry Batch Sizes

### LINKED ISSUE #1903: Munitions Crafter, Ammunition Crafting
In GitLab by @Velara on Nov 29, 2021, 17:25

At present when a Gunslinger with the munitions crafter feat adds an ammunition formula such as Rounds (Jazail) to their advanced alchemy section. An error message is displayed "Item is not compatible with crafting entry trait requirements, must match one of [["alchemical, bomb"], ["alchemical"]

Moreover, when crafting ammunition or Black Powder doses, they are treated like normal alchemical items and cost 1 reagent per 2, while level 0 ammunition should instead cost 1 reagent per 10 rounds.  Per the Munititions Crafter feat. "If you use a batch of infused reagents to create basic level-0 ammunition such as black powder cartridges or black powder doses, you produce 10 rounds of ammunition."

Awesome job on the crafting window, by the way. Love it.

### PR DESCRIPTION
**Pull request by ebuckle**
Alternate to: https://github.com/foundryvtt/pf2e/pull/3353

Resolves #1903

As title, instead of having fieldDiscovery being a trait, instead use a predicate fieldDiscoveryItems in the same way as craftableItems, allowing the problem of munitions crafter to be more cleanly solved.

The only pain with this is the field discovery ae-likes, since they're adding a RawPredicate postDerived and have to be validated in the crafting entry. Thoughts?

As I stated in the other PR, this still leaves a problem of Precious Munitions but that needs a wider fix to actually allow materials to be applied to ammunition.

### CODE REVIEW

**Comment by stwlam** [comment]
Rather than refactoring this, we should remove `fieldDiscovery` (and anything else alchemist-specific) from the rule element.

**Comment by ebuckle** [comment]
So yeah, my thinking to make this even more generic was to have an array of of predicate/batch sizes with each formula using the greatest valid batch size or the default.

**Comment by ebuckle** [comment]
Experiment with a batchSize RE to allow any predicate/batch size pairings to be adding to crafting entries.

Not too happy with this as an implementation. Ideally a crafting entry would have all the predicates by default, with new feats adding roll options to an actor rather than adding new predicates later.

Thoughts?

**Comment by ebuckle** [comment]
Crafting entries now have no mention of field discoveries, Alchemist/Munition crafter entries now have all the required predicate knowledge by default

**Comment by ebuckle** [comment]
Updated batch size field to use new predicates

**Comment by RobertBeilich** [comment]
Hi,
so what is the status on this?
#3510 was closed in favor of #3353, which in turn was closed in favor of this pr, but now this one is stale for a month.?

**Comment by stwlam** [comment]
At a glance, all the predicates are malformed.

**Comment by Far2Casual** [comment]
> Hi, so what is the status on this? #3510 was closed in favor of #3353, which in turn was closed in favor of this pr, but now this one is stale for a month.?

Thanks for bumping it up, after a year it gets boring for me to patch it up locally after every version :smile:

File: src/module/item/physical/document.ts [hunk 1]
```diff
@@ -161,6 +161,7 @@ abstract class PhysicalItemPF2e extends ItemPF2e {
             magical: this.isMagical,
             uninvested: this.isInvested === false,
             [`material:${this.material.precious?.type}`]: !!this.material.precious,
+            [`material-grade:${this.material.precious?.grade}`]: !!this.material.precious,
```
Comment (stwlam) [inline src/module/item/physical/document.ts:164]: Use `material:grade`

File: src/module/actor/character/crafting/entry.ts [hunk 2]
```diff
@@ -14,11 +14,11 @@ export class CraftingEntry implements Omit<CraftingEntryData, "parentItem"> {
...
     craftableItems: PredicatePF2e;
     maxSlots: number;
-    fieldDiscovery?: "bomb" | "elixir" | "mutagen" | "poison";
-    batchSize?: number;
-    fieldDiscoveryBatchSize?: number;
+    defaultBatchSize: number;
     maxItemLevel: number;
     parentItem: Embedded<ItemPF2e>;
+    batchSizes: BatchSizeData[];
+    batchSizeModifier: number;
```
Comment (stwlam) [inline src/module/actor/character/crafting/entry.ts:21]: Document all these

**Comment by ebuckle** [comment]
Good for another review.

I'll make a seperate contribution to the wiki detailing the re when I get some time

**Comment by ebuckle** [comment]
Is there anything else missing for this to be merged?

**Comment by RobertBeilich** [comment]
> Is there anything else missing for this to be merged?

@stwlam, any comment on this?

**Comment by Far2Casual** [comment]
@stwlam I'm sad to see this dropped. Ammunition Crafting (lvl 1 feat for Gunslinger) is still broken and requires me to patch my game locally every version for about a year and a half :(

**Pull request closed without merging**